stackslib/src/net/download/nakamoto.rs

```diff
@@ -0,0 +1,4283 @@
+    ///
+    /// Returns Ok(()) on sucess, and appends new tenures in the given reward cycle (`cur_rc`) to
+    /// `wanted_tenures`.
+    /// Returns Err(..) on DB errors.
+    pub(crate) fn update_wanted_tenures_for_reward_cycle(
+        cur_rc: u64,
+        tip: &BlockSnapshot,
+        sortdb: &SortitionDB,
+        wanted_tenures: &mut Vec<WantedTenure>,
+    ) -> Result<(), NetError> {
+        let highest_tenure_height = wanted_tenures.last().map(|wt| wt.burn_height).unwrap_or(0);
+
+        // careful -- need .saturating_sub(1) since this calculation puts the reward cycle start at
+        // block height 1 mod reward cycle len, but we really want 0 mod reward cycle len
+        let first_block_height = sortdb
+            .pox_constants
+            .reward_cycle_to_block_height(sortdb.first_block_height, cur_rc)
+            .saturating_sub(1)
+            .max(highest_tenure_height.saturating_add(1));
+
+        let last_block_height = sortdb
+            .pox_constants
+            .reward_cycle_to_block_height(sortdb.first_block_height, cur_rc.saturating_add(1))
+            .saturating_sub(1)
+            .min(tip.block_height.saturating_add(1));
+
+        if highest_tenure_height > last_block_height {
+            test_debug!(
+                "Will NOT update wanted tenures for reward cycle {}: {} > {}",
+                cur_rc,
+                highest_tenure_height,
+                last_block_height
+            );
+            return Ok(());
+        }
+
+        test_debug!(
+            "Update reward cycle sortitions between {} and {} (rc is {})",
+            first_block_height,
+            last_block_height,
+            cur_rc
+        );
+
+        // find all sortitions in this reward cycle
+        let ih = sortdb.index_handle(&tip.sortition_id);
+        let mut new_tenures =
+            Self::load_wanted_tenures(&ih, first_block_height, last_block_height)?;
+        wanted_tenures.append(&mut new_tenures);
+        Ok(())
+    }
+
+    /// Given the last-considered sortition tip and the current sortition tip, and a list of wanted
+    /// tenures loaded so far, load up any new wanted tenure data _in the same reward cycle_.  Used
+    /// during steady-state to load up new tenures after the sorittion DB advances.
+    ///
+    /// It may return zero tenures.
+    ///
+    /// Returns Ok(new-tenures) on success.
+    /// Returns Err(..) on error.
+    pub(crate) fn load_wanted_tenures_at_tip(
+        last_tip: Option<&BlockSnapshot>,
+        tip: &BlockSnapshot,
+        sortdb: &SortitionDB,
+        loaded_so_far: &[WantedTenure],
+    ) -> Result<Vec<WantedTenure>, NetError> {
+        let tip_rc = sortdb
+            .pox_constants
+            .block_height_to_reward_cycle(sortdb.first_block_height, tip.block_height)
+            .unwrap_or(0);
+
+        let first_block_height = if let Some(highest_wanted_tenure) = loaded_so_far.last() {
+            highest_wanted_tenure.burn_height + 1
```

Comment:
No sure, but we have a couple + 1 through here that I think should be saturating adds.

Reply:
I fixed the ones where we don't know the largest possible values of the addends